### ISSUE TITLE
Sparse matrix elements not scaled completely

### ISSUE THREAD

**Issue by phannebohm** [LLNL/sundials#256]
While working on a fix for #253, I found this part of `ScaleAddI` which is supposed to scale matrix A by c and add the identity:

https://github.com/LLNL/sundials/blob/7c318244182061ea482b449bdb5a14d4d3c71e57/src/sunmatrix/sparse/sunmatrix_sparse.c#L630-L637

It seems to me that only the left (top) part of A is scaled by c (if A is not square). Is this the desired behavior?

**Comment by drreynolds** [comment]
In SUNDIALS, `ScaleAddI` is only ever called when A is square.  In fact, sparse matrices are only ever used in SUNDIALS for square linear systems, but we wrote the module so that users could apply these tools internally within their codes for more general matrices.  

It is unclear what *should* be the desired behavior for the operation A <- cA + I, when A is not square, since it is mathematically illegal to add two matrices of different shape, and I is always square.  That said, the current implementation is indeed inconsistent.  It should either (a) return with an error when A is not square, or (b) scale all of A and then add as much of an identity to A as possible.  It looks like we went with (b) but there is a bug in that it forgets to scale the rest of A.  I'll ensure that this is resolved before our next release.
